src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java

```diff
@@ -59,159 +41,91 @@ public RepositoryAccessService(ParticipationAuthorizationCheckService participat
-        }
-        else if (!hasAccessToSubmission && hasPlagiarismComparison) {
-            throw new AccessForbiddenException();
-        }
-    }
+    private boolean hasWriteOrResetPermissions(StudentParticipation studentParticipation, boolean isLocked) {
+        boolean beforeDueDate = exerciseDateService.isBeforeDueDate(studentParticipation);
+        boolean isPracticeMode = studentParticipation.isPracticeMode();
 
-    /*
-     * The student can reset the repository only before and a tutor/instructor only after the due date has passed
-     */
-    private void checkAccessRepositoryForReset(ProgrammingExerciseParticipation programmingExerciseParticipation, boolean isStudent, ProgrammingExercise programmingExercise) {
-        boolean isOwner = true; // true for Solution- and TemplateProgrammingExerciseParticipation
-        if (programmingExerciseParticipation instanceof StudentParticipation studentParticipation) {
-            isOwner = authorizationCheckService.isOwnerOfParticipation(studentParticipation);
+        // The user has write or reset permissions if the participation is not locked and the due date has not passed yet.
+        if (beforeDueDate && !isLocked) {
+            return true;
         }
-        if (isStudent && programmingExerciseParticipation.isLocked()) {
-            throw new AccessForbiddenException();
-        }
-        // A tutor/instructor who is owner of the exercise should always be able to reset the repository
-        else if (!isStudent && !isOwner) {
-            // The user trying to reset the repository is at least a tutor in the course and the repository does not belong to them.
-            // This might be a tutor that resets the repository during assessment.
-            // If the student is still able to submit, don't allow for the tutor to reset the repository.
-
-            // For a regular course exercise, the participation must be locked.
-            if (programmingExercise.isCourseExercise() && !programmingExerciseParticipation.isLocked()) {
-                throw new AccessForbiddenException();
-            }
 
-            // For an exam exercise, the student must not be allowed to submit anymore.
-            // Retrieving the student via this type cast is safe because isOwner is false here which means that "programmingExerciseParticipation instanceof StudentParticipation"
-            // must have been true in the check above.
-            var optStudent = ((StudentParticipation) programmingExerciseParticipation).getStudent();
-            if (optStudent.isPresent() && programmingExercise.isExamExercise() && examSubmissionService.isAllowedToSubmitDuringExam(programmingExercise, optStudent.get(), false)) {
-                throw new AccessForbiddenException();
-            }
-        }
+        // The user has write or reset permissions if due date has passed, but the participation is in practice mode and not locked.
+        return !beforeDueDate && isPracticeMode && !isLocked;
```

Comment:
We can actually return `false` for `isLocked == true` even before the initialization of the first variables.